src/main/java/com/paloaltonetworks/osc/api/PANManagerSecurityGroupInterfaceApi.java

```diff
@@ -1,87 +1,263 @@
+            return null;
+        }
+
+        for (AddressEntry ae : addresses) {
+            Set<ManagerPolicyElement> currElements = extractPolicyElements(ae);
+            retVal.addManagerPolicyElements(currElements);
+        }
+
+        return retVal;
+    }
+
+    @Override
+    public String findSecurityGroupInterfaceByName(String name) throws Exception {
+        return null;
+    }
+
+    @Override
+    public List<? extends ManagerSecurityGroupInterfaceElement> listSecurityGroupInterfaces() throws Exception {
+        Map<String, ManagerSecurityGroupInterfaceElement> results = new HashMap<>();
+
+        for (AddressEntry ae : this.panClient.getAddressEntries(this.devGroup)) {
+
+            if (ae == null || ae.getTagNames() == null) {
+                continue;
+            }
+
+            String sgMgrId = ae.getTagNames().stream().filter(TagToSGIdUtil::isSGTag).findFirst().orElse(null);
+
+            if (sgMgrId != null) {
+                String sgInfcMgrId = combineToSGIntefaceId(sgMgrId, this.vs.getName());
+                PANSecurityGroupInterfaceElement res = (PANSecurityGroupInterfaceElement) results.get(sgInfcMgrId);
+                if (res == null) {
+                    res = new PANSecurityGroupInterfaceElement(sgInfcMgrId, sgInfcMgrId,
+                                                               new HashSet<>(), sgInfcMgrId, sgMgrId);
+                    results.put(sgInfcMgrId, res);
+                }
+
+                Set<ManagerPolicyElement> currElements = extractPolicyElements(ae);
+                res.addManagerPolicyElements(currElements);
+            } else {
+                LOG.error("Address object {} detected under {} with no security group", ae.getName(), this.devGroup);
+            }
+        }
+
+        return new ArrayList<>(results.values());
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private String doUpdateSecurityGroupInterface(SecurityGroupInterfaceElement sgiElement)
+            throws Exception {
+        String sgMgrId = sgiElement.getManagerSecurityGroupId();
+        String sgInterfaceId = combineToSGIntefaceId(sgMgrId, this.vs.getName());
+
+        LOG.info("Creating SGI {} for security group {}.", sgInterfaceId, sgMgrId);
+
+        List<AddressEntry> ipsOnMgr = this.panClient.fetchAddressesWithTag(sgMgrId, this.devGroup);
+
+        if (ipsOnMgr.isEmpty()) {
+            LOG.info("Security group {} on VS {} has no addresses. Cannot conform policies for SGI.", sgInterfaceId, this.vs.getName());
+            return null;
+        }
+
+        Set<String> policyTags = emptySet();
+        if (sgiElement.getManagerPolicyElements() != null) {
+            policyTags = sgiElement.getManagerPolicyElements().stream().filter(e -> e != null)
+                                   .map(e -> e.getName()).collect(toSet());
+        }
+
+        try {
+            for (AddressEntry entry : ipsOnMgr) {
+                List<String> tags = entry.getTagNames();

```

Comment:
I assume there is no setTags for entry API in PAN? then we can just call the same set method all the time.

In absence of that,naming the variables can be improved.

tags -> tagsToRemove
tmp -> tagsToAdd

tagsToAdd.removeAll(tagsToRemove);
tagsToRemove.removeAll(policyTags)